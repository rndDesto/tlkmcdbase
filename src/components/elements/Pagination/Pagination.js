import React from 'react';
import PropTypes from 'prop-types';
import queryString from 'querystring';
import { Link } from 'react-router-dom';
import styles from './styles.scoped.css';

export default function Pagination(props) {
  const { className, location, meta } = props;
  const { page, totalPage } = meta;
  const classes = [styles.root, className].filter(Boolean).join(' ');
  const toPrevPage = {
    page: page ? (parseInt(page) - 1) : 1,
  };
  const toNextPage = {
    page: page ? (parseInt(page) + 1) : 2,
  };
  return (
    <section className={classes}>
      <IconArrow
        disabled={page === 1}
        icon="../../../assets/ic-chevron-left.svg"
        location={location}
        title="Prev Page"
        to={toPrevPage}
      />
      <PageNumber location={location} meta={meta}  />
      <IconArrow
        disabled={page >= totalPage}
        icon="../../../assets/ic-chevron-right.svg"
        location={location}
        title="Next Page"
        to={toNextPage}
      />
    </section>
  );
}

Pagination.defaultProps = {
  className: '',
  location: {},
  meta: {},
};

Pagination.propTypes = {
  className: PropTypes.string,
  location: PropTypes.object,
  meta: PropTypes.object,
};


export function IconArrow({ disabled, title, icon, to, location }) {
  return (
    <Link className={styles.icon} disabled={disabled} to={getLink(to, location)}>
      <img src={icon} title={title} />
    </Link>
  );
}

IconArrow.defaultProps = {
  disabled:false,
  icon: '',
  location: {},
  title: '',
  to: {},
};

IconArrow.propTypes = {
  disabled:PropTypes.bool,
  icon: PropTypes.string,
  location: PropTypes.object,
  title: PropTypes.string,
  to: PropTypes.object,
};

export function PageNumber({ meta, location }) {
  const { page, totalPage } = meta;
  const totalPages = totalPage ? totalPage: '';
  const length = totalPages > 4 ? 4 : totalPages;
  const mainPages = Array.from(Array(length).keys()).map(i => {
    if (totalPages < 4 || page === 1) return i + 1;
    if (page === totalPages || page === totalPages - 1) return (totalPages - (3 - i));
    return page + i - 1;
  });
  const leftPages = totalPages > 5 && mainPages[0] - 1 > 0 ? [1, '...'] : [];
  const rightPages = totalPages > 5 && totalPages - mainPages[1] > 2 ? ['...', totalPages] : [];
  const pages = leftPages.concat(mainPages, rightPages);

  return (
    pages.map((item, key) => {
      const activePage = page === item && styles.active;
      const pageClasses = [styles['page-number'], activePage].filter(Boolean).join(' ');
      const disabled = typeof item !== 'number';
      const query = {
        page: item,
      };
      return (
        <Link className={pageClasses} disabled={disabled} key={key} to={getLink(query, location)}>
          {item}
        </Link>
      );
    })
  );
}

PageNumber.defaultProps = {
  location: {},
  meta: {}
};


PageNumber.propTypes = {
  location: PropTypes.object,
  meta: PropTypes.object
};

export function getLink(query, location) {
  return {
    pathname: location.pathname,
    search: `?${queryString.stringify({
      page: query.page
    })}`
  };
}
