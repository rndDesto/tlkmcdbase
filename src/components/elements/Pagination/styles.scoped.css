.root {
  display: flex;
  justify-content: flex-end;
  margin-top: 2.5rem;
}

.icon{
  margin-right: 0.8rem;
  margin-left: 0.8rem;
}

.active{
  background-color: var(--primary-color);
  border-radius: 0.1875rem;
  padding: 0.1875rem, 0.5rem;
  color: white;
}

.page-number {
  align-self: center;
  align-items: center;
  border-radius: 0.1875rem;
  flex: none;
  margin: 0 0.25rem;
  text-align: center;
  width: 1.5rem;
}