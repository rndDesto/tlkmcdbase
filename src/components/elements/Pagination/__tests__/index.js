import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Pagination, { IconArrow, PageNumber, getLink } from '../Pagination';

jest.mock('querystring', () => ({
  stringify: () => 'query',
}));

jest.mock('react-router-dom', () => ({
  Link: function link(props) { return <a {...props} />; },
}));

describe('src/components/elements/Pagination', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Pagination />);
    expect(tree).toMatchSnapshot();

    const tree2 = Pagination({ ...Pagination.defaultProps, meta: { page: 2 } });
    expect(tree2.props.children[0].props.to.page).toBe(1);
  });

  test('IconArrow', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<IconArrow />);
    expect(tree).toMatchSnapshot();
  });

  test('PageNumber', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<PageNumber />);
    expect(tree).toMatchSnapshot();

    const meta1 = { page: 1, totalPage: 1 };
    PageNumber({ ...PageNumber.defaultProps, meta: meta1 });

    const meta2 = { page: 5, totalPage: 6 };
    const tree2 = PageNumber({ ...PageNumber.defaultProps, meta: meta2 });
    expect(tree2[0].props.children).toBe(1);
    expect(tree2[1].props.children).toBe('...');

    const meta3 = { page: 2, totalPage: 6 };
    const tree3 = PageNumber({ ...PageNumber.defaultProps, meta: meta3 });
    expect(tree3[4].props.children).toBe('...');
  });

  test('getLink', () => {
    const res = getLink({}, { pathname: 'path' });
    expect(res.pathname).toBe('path');
    expect(res.search).toBe('?query');
  });
});
