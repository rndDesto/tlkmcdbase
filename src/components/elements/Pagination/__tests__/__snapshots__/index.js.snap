// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/elements/Pagination IconArrow 1`] = `
<link
  disabled={false}
  to={
    Object {
      "pathname": undefined,
      "search": "?query",
    }
  }
>
  <img
    src=""
    title=""
  />
</link>
`;

exports[`src/components/elements/Pagination PageNumber 1`] = `
Array [
  <link
    className=""
    disabled={false}
    to={
      Object {
        "pathname": undefined,
        "search": "?query",
      }
    }
  >
    1
  </link>,
]
`;

exports[`src/components/elements/Pagination render 1`] = `
<section
  className=""
>
  <IconArrow
    disabled={false}
    icon="../../../assets/ic-chevron-left.svg"
    location={Object {}}
    title="Prev Page"
    to={
      Object {
        "page": 1,
      }
    }
  />
  <PageNumber
    location={Object {}}
    meta={Object {}}
  />
  <IconArrow
    disabled={false}
    icon="../../../assets/ic-chevron-right.svg"
    location={Object {}}
    title="Next Page"
    to={
      Object {
        "page": 2,
      }
    }
  />
</section>
`;
